### ISSUE TITLE
Color management needs to improve

### ISSUE THREAD

**Issue by fitu996** [nomacs/nomacs#459]
**Is your feature request related to a problem? Please describe.**
There has been an old unsolved issue like #386 #370 #204 because Qt lacks the support of the color profile.

**Describe the solution you'd like**
I think there might be a walkaround that converts all color space to standard sRGB before displaying the image \(the image processing module in nomacs will still see the untouched image\). There is some libraries like [LittleCMS](http://www.littlecms.com/index.html) or [Magick++](https://www.imagemagick.org/Magick++/Image++.html) will correctly handle the color conversion.

**Comment by diemmarkus** [comment]
the problem here is that Qt converts i.e. CMYK jpgs in the loader to RGB. Hence we would have to change the plugin itself, which is bad for future updates...

**Comment by fitu996** [comment]
Will it be a bad idea that load image using magick++/OpenCV and then convert it to QImage or something that used in nomacs?

Where is the code that nomacs open jpeg/tiff files that may contains color profiles? Sorry, but I can't find it. Maybe I can help you solve this issue.

**Comment by diemmarkus** [comment]
using magick is not too good (new dependency)... however, `OpenCV` is there anyhow and we use it for multi-page tiff support already. so there could be a way. A few things we need to check before would be:
- Does OpenCV support ICC profiles?
- converting `cv::Mat`to `QImage` has a bit of a performance offset (that is acceptable)

> Where is the code that nomacs open jpeg/tiff files that may contains color profiles?

the color conversion code is external (in the Qt JPG plugin). that's why changing it is not straight forward...

**Comment by fitu996** [comment]
> Does OpenCV support ICC profiles?

Unfortunately, no. OpenCV is focused on computer vision \(CV\), but color management is not CV.

> using magick is not too good (new dependency)
- Yes, magick is a big library. It contains so many compile options to enable its functionally.
- But, it does have so many useful features. It focused on image processing, which can be regarded as a supplement to OpenCV.
- It can handle massive amounts of image format and support more options when saving an image \(compared to Qt\). It can quickly solve all feature requests like "Please add support for xxx image format".
- So, I believe it is reasonable that an "Image Viewer/Editor" have such a dependency. \(OpenCV is somehow a bit library, right?\)

Magick can be considered as an interface. It has a lot of implementations such as [ImageMagick](https://imagemagick.org/index.php) and [GraphicsMagick](http://www.graphicsmagick.org/).

For a lightweight dependency, [LittleCMS](http://www.littlecms.com/index.html) can be a good choice. We can use OpenCV to open the image, then use LittleCMS to convert it.

Another solution: Is it possible to make a plugin called "ImageMagick image opener" that open images using ImageMagick then convert it to QImage using a "correct way"? This will not add new dependencies to the main program.

**Comment by diemmarkus** [comment]
from my point of view, we don't want magick dependency in the main repo due to portability... so there are basically two options you named left:

> For a lightweight dependency, [LittleCMS](http://www.littlecms.com/index.html) can be a good choice. We can use OpenCV to open the image, then use LittleCMS to convert it.

advantage: little extra dependencies, probably straight forward to implement. This could also be solved using a plugin.

> Another solution: Is it possible to make a plugin called "ImageMagick image opener" that open images using ImageMagick then convert it to QImage using a "correct way"? This will not add new dependencies to the main program.

+ as you said, we solve a bunch of loading issues.
- again: heavy dependency, more work to implement

In general, it's best to develop a [QImageIOPlugin](https://doc.qt.io/qt-5/qimageioplugin.html). A good sample is the [HEIF Plugin](https://github.com/jakar/qt-heif-image-plugin) which we use for loading heif/heic files because it also wraps around an existing loading library. Advantage of a Qt plugin would be that other Qt based programs would benefit as well.

Would you be interested in developing such a plugin? I could support you to some extent (given limited time resources).

**Comment by fitu996** [comment]
Sure. Give me some time to get familiar with QImageIOPlugin.

So far this method has some drawbacks:
1. As far as I know, jpg&tiff \(maybe more\) has support for embedded color profile. HEIF has support for using a standard color space beside RGB. These image format is currently supported by nomacs, some of them with write support. Making such a QImageIOPlugin will break the write support.
2. In the future, there may be more and more image format supports different color spaces. Does this method need to create a QImageIOPlugin for every image container?
3. Some color space such as HDR color spaces have the ability to present a color which cannot be presented by sRGB. So converting color space for these images will use some lossy algorithm, which is impossible to revert. Giving such images write support is not a good idea before Nomacs gain native support for these color spaces. \(OpenCV have this support but Nomacs use QImage everywhere\)

I prefer a more general way to solve the color management problem because it is not a problem caused by a specific image container.

**Comment by fitu996** [comment]
It seems that Nomacs handle every image as linear RGB888 format \(even some HEIF image uses YUV420p16 format, it will be converted to RGB\). So completely solve the color problem cannot be done in an easy way. It requires using Qt for UI only and OpenCV for all other image processing. OpenCV will give the native support for RGB, RGBA, YUV, CYMK, etc. format in various bit-depth. And LittleCMS will perform the color management staff and helps to converts these formats to QImage.

All the approaches we have talked about before are walkaround, not prefect solution. So if it is too expensive to solve this problem perfectly, I hope the walkaround is something more general.

**Comment by fitu996** [comment]
So, it will be awesome that Nomacs 4.0 have native support for image format other than RGB and color space other than linear. Not just a walkaround.

**Comment by diemmarkus** [comment]
ok, that's a lot of input : ) please let me know if I missed a point.

To make one thing clear before discussing others: nomacs is an image *viewer*. So beside some editing capabilities that we added, its main purpose is to display images on a (typically RGB) monitor fast. That is why I don't see converting images to sRGB as a hack or work around. Handling different color spaces in the memory would:
- consume more RAM (i.e. for CMYK or 16-bit)
- slow down displaying (in the end we need to convert to the monitor's color space at some point)
- dramatically complicate the code -> introduce bugs

that's why I would have thought of an image plugin that convert's various color spaces to the monitor's sRGB using `QImage`. Saving images to color spaces other than RGB (which is not trivial) is in my opinion a task for image editors (i.e. Photoshop/GIMP).

>     1. Making such a QImageIOPlugin will break the write support.

you can specify what formats your plugin can read or write [QImageIOPlugin::Capability](https://doc.qt.io/qt-5/qimageioplugin.html#Capability-enum). If you don't support write of i.e. jpg, Qt will fallback to it's default jpg plugin.

>     2\. create a QImageIOPlugin for every image container?

Typically every format has it's own plugin. But you can also support multiple formats with one plugin. Do you plan to rely on magick for loading?

> I prefer a more general way to solve the color management problem because it is not a problem caused by a specific image container.

We currently load all images (except for raw and tiff) through QImageIOPlugins so, there is no way other than replacing these for color conversion (because they already convert images to RGB).

**Comment by fitu996** [comment]
Oh, I see your opinion `Nomacs is an image viewer`. Since it is not focusing on editing, converting everything to RGB is reasonable. But it is better to give users some warning like "You are saving a picture as RGB color space, while the source is CYMK".

If I developed a QImageIOPlugin, how do Nomacs decide when to use my plugin and when use the QT's built-in? \(eg. Will Nomacs treats normal jpeg and jpeg contains color profiles as different image type?\)

The more general way in my imagination is to change the image loading logic to:
1. Load the raw image data using OpenCV or something else
2. Convert colorspace to RGB
3. Convert the image to QImage

But, simply developing a QImageIOPlugin is good enough. Comparing it to my solution, creating a QImageIOPlugin can cause program coupling to be a little serious and reduce the scalability.

There is a new monitor type called HDR monitor. It is still RGB but it will display many more colors. Although Qt does not currently support this monitor, it will be a problem if Nomacs do not support high bit-depth \(such as 16bit\) RGB picture in the future.

**Comment by diemmarkus** [comment]
sorry for the late reply here...


> how do Nomacs decide when to use my plugin and when use the QT's built-in?

this is not possible according to the [docs](https://doc.qt.io/qt-5/qimageioplugin.html#details):
````
If several plugins support the same capability, Qt will select one arbitrarily.
````

> is to change the image loading logic to

we could also pursue this idea. Here I would first check if there is a CMYK image about to be loaded & only then trigger the above mentioned method (because it's heavier than the default loading). Again: if you load images using `something else`, the whole method needs to be a plugin (we can create our own interface to circumvent the issues with QImageIOPlugin) so that no additional dependencies are added to the framework.

> There is a new monitor type called HDR monitor.

I know - but we cannot support HDR monitor's without Qt (we'd have to write our own rendering engine - which is beyond the scope of nomacs). But I just noticed that at least QImage ``QImage::Format_RGBA64`` now supports 16-bit (which is also nice for editing).

**Comment by fitu996** [comment]
So, your solution is to write an image loading plugin using Nomacs' plugin interface and then write the logic to decide whether to use it? It would be great if there is an interface which let the plugin to decide when to use itself.

Or, your solution is to write a QImageIOPlugin and take responsibility for opening all image formats? It seems like it will slow down the Nomacs.

Or, your solution is to write a QImageIOPlugin and take responsibility for opening all jpeg and tiff images? It still has some unneeded performance impact.

**Comment by diemmarkus** [comment]
> So, your solution is to write an image loading plugin using Nomacs' plugin interface and then write the logic to decide whether to use it? It would be great if there is an interface which let the plugin to decide when to use itself.

I guess that's the best solution. here are two thoughts considering who will decide what images to open:
- at loading stage, nomacs has the full metadata already extracted (so checking if an image is cmyk costs nothing).
- we can for sure add a Qt like call to the plugin to see if it accepts the image at hand (but I would only call it if the image has a colorspace other than RGB for what we have right now is fine there...)

**Comment by fitu996** [comment]
> at loading stage, nomacs has the full metadata already extracted (so checking if an image is cmyk costs nothing).

It sounds good enough. How can I help you?

**Comment by zohaibaliG** [comment]
Would love to see improvements in this regard. Otherwise it's a great alternative to the any other apps I've tried so far : )

**Comment by scrubbbbs** [comment]
There seems to be a path to doing this now in Qt 6.8+
- We now have QColorSpace which is attached to QImage, and can do conversions and load ICC profiles
- Some image loaders are already providing the native pixel format and/or colorspace so we can display them correctly:
  - CMYK tiff and jpg
  - JXR  files with Rec.2020 profile
  - HEIC files with DisplayP3 profile

**Comment by fililip** [comment]
Would it also be possible to implement the Wayland color management protocol?

**Comment by scrubbbbs** [comment]
> Would it also be possible to implement the Wayland color management protocol?

I just learned that Wayland compositor wants to implement color management itself, unlike X11 where it is up to applications. Which means we probably just convert everything to sRGB if we detect Wayland -- or maybe do this by default for all platforms. Eventually we might have a way to set the color profile of a subwindow/surface to match the image and advertise this to Wayland, but the support in Wayland is not there yet FWICT.

**Comment by fililip** [comment]
The protocol is shared between compositors which implement this functionality (so that everything is unified); an excellent example of this is KWin, where HDR/color management just works with the protocol: https://wayland.app/protocols/color-management-v1.

Here is an example application that uses it in a Qt project: https://invent.kde.org/zamundaaa/colortest

But yes, currently if an application does not use this protocol on Wayland, sRGB is automatically assumed. It would be great if we could display images such as [these](https://invent.kde.org/zamundaaa/colortest/-/blob/master/src/contents/images/HDR%20city.png?ref_type=heads) in nomacs. Right now one has to use mpv with either the Vulkan or dmabuf-wayland backend to do it, which works great, but that is not ideal, as that's a video player, not an image viewer.

Krita also has a WIP branch: https://invent.kde.org/graphics/krita/-/merge_requests/2457

**Comment by Orum** [comment]
I noticed that the latest release (3.22.0) claims to have color management support, thanks to #1454.  After reading through the PR, it seems like this is still rather limited support (e.g. it is limited to outputting in sRGB, which is less than ideal on WG displays).

However, I'm not sure it's even working properly, as WG images are _not_ being tone mapped to sRGB.  AFAICT they are simply clipped, which I'd hardly call "color management".  This is very easy to see with some of the [WebKit wide gamut test images](https://webkit.org/blog-files/color-gamut/), particularly the [WebKit logo image](https://webkit.org/blog-files/color-gamut/Webkit-logo-P3.png) which just appears as a red square when converting from P3 to 709 primaries without being tone mapped (i.e. just clipping).

Is there any chance we'll get proper tone mapping in the future, and ideally a way to output in wider color spaces than just sRGB?

**Comment by scrubbbbs** [comment]
> WG images are _not_ being tone mapped to sRGB

Thanks for taking a look at this. We aren't trying to detect the display profile in 3.22.0, so sRGB is the default profile. You may select one of the preset output profiles in `Settings/Display/Color Management`, or load your own .icc file specific to your monitor.

Note if you are using Linux/Wayland this won't work (see previous comment) as it requires specific integration with Wayland. Setting the profile to sRGB is the only option there for the moment.

**Comment by Orum** [comment]
> Note if you are using Linux/Wayland this won't work (see previous comment) as it requires specific integration with Wayland. Setting the profile to sRGB is the only option there for the moment.

Yes, as I'm using kwin/Wayland, that was my understanding, which is why I've left it set as 'sRGB' in the configuration.  The issue is how it converts wider gamut images into sRGB's limited gamut, clipping them instead of mapping/scaling them.

**Comment by scrubbbbs** [comment]
> The issue is how it converts wider gamut images into sRGB's limited gamut, clipping them instead of mapping/scaling them.

I think you are talking about tone mapping, which is a separate issue from color management. We do not support tone mapping yet, which includes ICC Perceptual Intent.

- Color management is to make colors look *correct* on different displays. Clipping is not incorrect behavior; the target profile (in this case, sRGB) cannot display that color so there is nothing else to be done but clip to the nearest color.

- Tone mapping is to make images look *subjectively better* when the display cannot show the full range of colors, by applying some sort of perceptual model (there are quite a few for HDR, for SDR it's much easier). This would be a nice addition to nomacs.

**Comment by Orum** [comment]
>     * Color management is to make colors look _correct_ on different displays. Clipping is not incorrect behavior; the target profile (in this case, sRGB) cannot display that color so there is nothing else to be done but clip to the nearest color.

But that's just it--they don't look correct, even when all the colors in the image are in the limited gamut of sRGB.

For example, if I open an image in mpv, which I've configured like so:
```
target-colorspace-hint=no
target-prim=bt.2020
target-trc=bt.1886
```

The image looks correct on my display (which is configured the same way, as is kwin), whether the image is just your bog standard sRGB (709 primaries/1886 transfer), or a wide gamut image.  However, in nomacs, they always look different (basically oversaturated) compared to mpv.

**Edit:** Out of curiosity I tried setting the color management to the same ICC profile (bt 2020) that I use in my display and compositor.  This seems like it should be the wrong thing to do based on what you said earlier ("Setting the profile to sRGB is the only option there for the moment."), but it does seem to get the correct output, or at least identical to mpv's output, for _some_ images, but not all.

It appears that sRGB images now look the same whether viewed in mpv or in nomacs using the ICC profile, but wide gamut images still look wrong in nomacs (they look brighter and more washed out than mpv's output, at least in the images I've tested).

**Comment by scrubbbbs** [comment]
> This seems like it should be the wrong thing to do based on what you said earlier ("Setting the profile to sRGB is the only option there for the moment."),

Yeah it definitely seems wrong. Maybe there is a kwin setting that changes this behavior, or you are actually running on X11.

> it does seem to get the correct output, or at least identical to mpv's output, for _some_ images, but not all.

I did a quick test on my sRGB display (X11), and I need this to get a match between nomacs and mpv. As I said we do not do tone mapping/gamut mapping yet, so those must be disabled in mpv as well.

```
target-colorspace-hint=no
target-prim=bt.709
target-trc=srgb
tone-mapping=clip
gamut-mapping-mode=clip
```

Here is the result. 
Webkit logo:

<img width="1291" height="717" alt="Image" src="https://github.com/user-attachments/assets/ee04dfe9-fbe4-4eff-a58d-09e09a10acb6" />

HDR image (still have some issues here, but close):

<img width="1291" height="717" alt="Image" src="https://github.com/user-attachments/assets/db4b0d36-a578-417f-9531-cab00713f7b4" />

Rec2020 test file:

<img width="1459" height="670" alt="Image" src="https://github.com/user-attachments/assets/130a55d2-1221-4c83-9454-0ea9537f61ba" />

Another HDR image with linear color space, this is totally off in nomacs vs mpv. But looks the same in gwenview and gimp so I'm not sure this is a bug.

<img width="1161" height="518" alt="Image" src="https://github.com/user-attachments/assets/acc04656-20f2-4ef1-bdd9-7d0c04e15b4c" />

**Comment by Orum** [comment]
> Yeah it definitely seems wrong. Maybe there is a kwin setting that changes this behavior, or you are actually running on X11.

Fairly certain it's wayland as that's selected when logging in and XDG_SESSION_TYPE is set to 'wayland'.

> I did a quick test on my sRGB display (X11), and I need this to get a match between nomacs and mpv. As I said we do not do tone mapping/gamut mapping yet, so those must be disabled in mpv as well.

Interesting that the wide gamut images look the same for you, but as our hardware and software setup is so different, I'm not sure it means anything for my situation.  I'll leave nomacs set to bt.2020 for now, because at least it appears to show sRGB images correctly (and some others I think are accurate, like P3), and just be careful not to open anything tagged as bt.2020 in it.
